torch/csrc/jit/mobile/import.cpp

```diff
@@ -424,17 +424,53 @@ void BytecodeDeserializer::parseMethods(
 
     static const c10::QualifiedName classPrefix = "__torch__.torch.classes";
     for (const auto& t : types_list) {
-      c10::QualifiedName qn(t.toStringRef());
-      if (classPrefix.isPrefixOf(qn)) {
-        auto classType = getCustomClass(qn.qualifiedName());
-        TORCH_CHECK(
-            classType,
-            "The implementation of class ",
-            qn.qualifiedName(),
-            " cannot be found.");
-        function->append_type(classType);
+      if (t.isString()) {
+        c10::QualifiedName qn(t.toStringRef());
+        if (classPrefix.isPrefixOf(qn)) {
+          auto classType = getCustomClass(qn.qualifiedName());
+          TORCH_CHECK(
+              classType,
+              "The implementation of class ",
+              qn.qualifiedName(),
+              " cannot be found.");
+          function->append_type(classType);
+        } else {
+          function->append_type(c10::parseType(t.toStringRef()));
+        }
       } else {
-        function->append_type(c10::parseType(t.toStringRef()));
+        //  Example NamedTuple type structure:
+        //  ('__torch__.A.B.CType',z
+        //   ('NamedTuple',
+        //     ('id_list_features', 'Dict[int, Tensor]'),
+        //     ('label', 'Tuple[Tensor, Tensor]'),
+        //     ('weight', 'Tuple[Tensor, Tensor]'),
+        //     ('id_score_list_features', 'Dict[int, Tensor]')))),
+        auto tuple_difinition = t.toTuple()->elements();
+        // get type name: '__torch__.A.B.CType'
+        const std::string name = tuple_difinition[0].toString()->string();
+        // get namedtuple definition
+        std::vector<IValue> name_type_pairs =
+            tuple_difinition[1].toTuple()->elements();
+
+        at::QualifiedName qualified_name = at::QualifiedName(name);
+        std::vector<std::string> field_names;
+        std::vector<TypePtr> field_types;
+
+        // Find all type names and it's corresponding type
+        for (auto const& name_type_pair :
+             name_type_pairs[1].toTuple()->elements()) {
+          std::vector<IValue> name_type_vector =
+              name_type_pair.toTuple()->elements();
+          std::string field_name = name_type_vector[0].toString()->string();
+          TypePtr field_type =
+              c10::parseType(name_type_vector[1].toString()->string());
+          field_names.emplace_back(field_name);
+          field_types.emplace_back(field_type);
+        }
+        // Create the NamedTuple type after reading from the tuple, and add it to function
```

Comment:
And then this is me just not understanding how mobile deserialization works, but its not intuitive to me that we are adding the NamedTuple type fo the function its called from instead of say the module as a member or something. Can you maybe elaborate on this?

Reply:
A type is associated with the byte code of that function. For example, when running LIST_CONSTRUCT of a function, it fetches the type from the type table: `const auto& type = code_->types_[inst.X]->expectRef<at::ListType>();`